<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        async function a(){
            console.log(1);

            setTimeout(()=>{
                console.log(2);
            },0)

            await new Promise((resolve)=>{
                console.log(3)
                resolve();
            })

            console.log(5)

            // return 123;
        }

        // function a(){
        //     console.log(1);

        //     setTimeout(()=>{
        //         console.log(2);
        //     },0)

        //     return new Promise((resolve)=>{
        //         console.log(3)
        //         resolve();
        //     })
        //     .then(()=>{
        //         console.log(5)
        //         return 123;
        //     })

        //     // console.log(5)
        // }

        async function b(){
            console.log(6);
            await a();
            console.log(7);
        }
        b();

        /*
            代码执行顺序:执行主线程代码->清空微任务队列->执行宏任务队列中的宏任务
            宏任务:setTimeout
            微任务:then
        
            async和await他们是promise和.then的语法糖
                其实就是await之后的代码都会被放在.then中执行

            async函数中无论是否有写return,返回值都一定是promise对象
                如何控制返回的promise对象的状态
                    变为成功
                        只要async函数中所有的代码都成功执行结束,那么返回的promise对象就会变为成功状态

                    变为失败
                        如果async函数中执行代码出现错误,那么返回的promise对象就会变为失败状态
                
        */
    </script>
</body>
</html>